### DISCUSSION TITLE
Wakamonth - hours reporting tool

### DISCUSSION THREAD

**Discussion by ghost** [muety/wakapi#580] [Show and tell]
Generating declarable hours reports using Wakapi with "Wakamonth";  a simple script that uses the Wakapi API to generate an Excel sheet that lists worked hours grouped per branch. Not sure if it's any good yet, but it is a start :) See https://github.com/bitstillery/wakamonth

Any thoughts about the reliability, improvements or anything else? Somehow it still has a lot of hours grouped on Unknown branch. Any idea how to debug this properly? I'm using the VScode plugin and the browser plugin (Firefox)

**Comment by muety** [comment]
Amazing little project, thanks for sharing! :clap: It's very cool to see tooling evolve around Wakapi.

Regarding your question:
> Somehow it still has a lot of hours grouped on Unknown branch. Any idea how to debug this properly?

When tracking browsing data using the WakaTime Firefox / Chrome plugin, the `branch` and `language` fields will be empty (-> "unknown") as there is no such thing as a "branch" for a website obviously. 

I tried the tool and it works very well! A few remarks from my side (knowingly, that the tool is, of course, still work in progress):
- Probably planned, but you'll likely want to pass the target project as a command-line argument as well
- From my understanding, you're currently rounding to hour-precision. Would be cool to be able to specify the "granularity" of reporting, e.g. half- or quarter-hourly
- I'd suggest using the WakaTime compat endpoint (see [here](https://wakatime.com/developers/#summaries), [here](https://wakapi.dev/swagger-ui/swagger-ui/index.html#/wakatime/get-wakatime-summaries)). It exposes the same data, but in a slightly different format. When querying against that endpoint, you'll implicitly get support for WakaTime (not only Wakapi) as well, which makes you're tool even more useful. 

And, of course, as a disclaimer: coding time tracked by Wakapi is only a rough proxy or indicator of the actual amount of work spent on a project / on a feature. It's only an estimate and definitely not legally binding, so people should keep that in mind when using it for billing reports / invoices.

**Comment by ghost** [comment]
Hi @muety,

Thanks for the feedback and suggestions! I added the target project as a config directive (forgot about that one) and added a precision parameter (60/30/15) for granularity. The disclaimer was added to the Readme, so people are warned up-front. I tried to use the compat endpoint, but Wakapi crashed after I called the endpoint like:
```
https://wakapidomain.org/api/compat/wakatime/v1/users/jvanveen/summaries?end=2023-12-31&start=2023-12-01&project=manhattan
```

About the unknown hours; I thought the plugin would only register whitelisted sites; e.g.
```
Logging Style: Only whitelisted sites
http://localhost:9000@@myproject
http://localhost:9001@@myproject
```
Shouldn't it log only hours on project/branch, based on <<LAST_BRANCH>>/<<LAST_PROJECT>>? 
I forgot to mention I'm also using the zsh plugin for Wakatime, which may also cause some unknown time.

**Comment by muety** [comment]
> but Wakapi crashed after I called the endpoint

That definitely shouldn't be happening. Could you please post the error message with which the server had crashed? 

> Shouldn't it log only hours on project/branch, based on <<LAST_BRANCH>>/<<LAST_PROJECT>>?

Yes, you're right! The branch field of a heartbeat sent from the browser plugin will be set to the last branch of that heartbeat's project. This requires a project to be explicitly specified for the site (using the `@@`), because Wakapi does intentionally not support the `<<LAST_PROJECT>>` feature (see our previous [discussion](https://github.com/muety/wakapi/discussions/532#discussioncomment-7195855)). So effectively, you'll only get a non-_unknown_ branch for site that you have whitelisted and defined a project mapping for.

Could you possibly have a look into your database's `heartbeats` table and see where most unknown branches (`branch` column being set to either empty string (`""`), NULL or `"<<LAST_BRANCH>>"`) come from? Would be interesting to know if they result from a bug in Wakapi or if it's the editor plugin's fault.

Also, which version of the browser plugin are you using? 

Either way, thanks a lot your efforts with _wakamonth_!

**Reply by ghost** [reply]
The logs show this exception when calling the endpoint:
```bash
Dec 31 13:27:25 wakapi[2486245]: panic: runtime error: index out of range [0] with length 0
Dec 31 13:27:25 wakapi[2486245]: goroutine 329 [running]:
Dec 31 13:27:25 wakapi[2486245]: github.com/muety/wakapi/models/compat/wakatime/v1.newDataFrom.func7(0xc0010ec2d0)
Dec 31 13:27:25 wakapi[2486245]:         /opt/wakapi/models/compat/wakatime/v1/summaries.go:189 +0x3d3
Dec 31 13:27:25 wakapi[2486245]: created by github.com/muety/wakapi/models/compat/wakatime/v1.newDataFrom
Dec 31 13:27:25 wakapi[2486245]:         /opt/wakapi/models/compat/wakatime/v1/summaries.go:186 +0xdbb
Dec 31 13:27:25 systemd[1]: wakapi.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```
Does this help, or should I increase the logging verbosity (if possible)?

Looking at the heartbeats table, the terminal plugin (sobolevn/wakatime-zsh-plugin) produces only empty branch references. I'll remove it for now to keep the branch data a bit cleaner. 

The browser plugin is less decisive. It has ~7500 entries without a branch, and about 11500 with a branch. The entries with a branch end with id 72445; the ones without a branch end with id 72512. It seems to suggest that recent heartbeats from that plugin are not consistently assigned a branch after I updated browser-wakatime to the <<LAST_BRANCH>> version (https://github.com/wakatime/browser-wakatime/issues/232)  I'm currently using version 3.0.21 (Last updated November 22, 2023). The commit is tagged on 3.0.20 and 3.0.21, so it should be active I guess. The browser heartbeats that do have a branch seem perfectly fine, where <<LAST_BRANCH>> is the branch I worked on at that moment.

Example entries look like:
```bash
# Vscode (branch set)
| entity: /path/to/file | type: file | category: coding | project: wakamonth | branch: main | language : JavaScript | editor: Vscode |
# Browser entry (branch set)
| entity: http://localhost:9001 | type: url | category: - | project: wakamonth | branch: main | language : - | editor: Firefox |

# Browser entry (empty branch)
| entity: http://localhost:9001 | type: url | category: - | project: wakamonth | branch: - | language : - | editor: Firefox |
# Terminal entry (empty branch)
| entity: ls | type: app | category: coding | project: wakamonth | branch: - | language: sh | editor: -
```

Ps. Would it be feasible to set the language coming from browser-wakatime heartbeats? In my case, this would be scss, css or html, but most likely scss(I spend most time styling in the browser).

**Reply by muety** [reply]
> Does this help, or should I increase the logging verbosity (if possible)?

That helps, will fix this! 

> Would it be feasible to set the language coming from browser-wakatime heartbeats?

Can you elaborate a bit more on how you imagine this to work? Are you thinking of something like a `<<LAST_LANGUAGE>>` functionality, i.e. setting the heartbeat's language to whatever was the last language you coded in in your IDE?

Regarding the empty branch heartbeats: could you possibly post an excerpt (or screenshot or so) of your `heartbeats` table where there is a heartbeat that should have a branch set, but doesn't, and in addition please also include the preceding and following ~ 10 heartbeats or so? I think this would help me debugging the issue.

**Reply by muety** [reply]
Also, what Wakapi version are you using?

**Reply by muety** [reply]
Regarding the missing branches from the ZSH plugin, I filed https://github.com/sobolevn/wakatime-zsh-plugin/issues/22.

**Reply by ghost** [reply]
> Can you elaborate a bit more on how you imagine this to work? Are you thinking of something like a <<LAST_LANGUAGE>> functionality, i.e. setting the heartbeat's language to whatever was the last language you coded in in your IDE?

Sure, I think LAST_LANGUAGE would be too inaccurate though. The last language (jsx/js/python in my case) in Vscode is probably different than the one I would like to refer to from the browser plugin(scss, css, or html). Not sure if this makes sense, but could a language suggestion maybe be part of the same convention that is being used to target a project? e.g.

```
http://localhost:9000@@myproject
http://localhost:9000@@myproject/language
```

> Also, what Wakapi version are you using?

It's the master branch at commit bc8e7f1234d313364120a4051d9c5fe0b6a0146d (tag 2.9.2)

**Reply by muety** [reply]
> [...] but could a language suggestion maybe be part of the same convention that is being used to target a project?

That's something you'd have to request upstream in the wakatime-cli, I think.

Could you please upgrade your Wakapi to the latest version? I believe the bug you reported had already been fixed in the meantime.

**Reply by ghost** [reply]
I just updated by switching tags from 2.9.2 to 2.10.2 and restarted. Somehow I can't seem to login anymore

**Reply by muety** [reply]
How did you do that update? And are you running Wakapi natively or in Docker?

**Reply by ghost** [reply]
Sorry, I rushed it a bit. I'm running it outside of docker and updated like:
```
git fetch
git checkout 2.10.2
go build -o wakapi
/opt/wakapi/wakapi -config /etc/wakapi.yml
```
The service starts fine, but I don't have smtp installed.  I tried to enter a gmail account smtp in the config, but the email sending times out. Running behind a Nginx proxy. Could this work?
```
  smtp:
    host: smtp.gmail.com
    port: 587
    username: mygmail@gmail.com
    password: <somepassword>
    tls: true
```

Or is there maybe another commandline escape-hatch to reset the password? 
When running reports on the data using the api key, there are no results on branches anymore. Does Wakapi need to regenerate stats from the heartbeats after an upgrade?

**Reply by muety** [reply]
Port 587 is usually used for `STARTTLS`, in which case you'd set `tls: false`. You shouldn't have to reset your password or regenerate summaries after an update. Did you perhaps accidentally delete your database or so?

**Reply by muety** [reply]
Those parameters shouldn't be irrelevant for you, I think. But maybe check your browser console for errors.

**Reply by ghost** [reply]
The reporting was empty because we just changed years, and I didn't add a year parameter yet. Best wishes for 2024 by the way :)
Gladly, the reporting data still works using the API key; only the frontend login usiny my credentials doesn't. The user has a password set($argon...), and the login itself seems to be fine (just returns a 401 as expected)
![Screenshot_20240102_201418](https://github.com/muety/wakapi/assets/48799939/5c53d983-e85b-4627-8e31-5a5f882e5b1d)

**Reply by muety** [reply]
Very strange. Did you change your salt (`security.password_salt` / `WAKAPI_PASSWORD_SALT`) after the update?

**Reply by ghost** [reply]
Ok, problem solved. I just had to restart the systemd service, and then the login worked as expected. Maybe a write permission error somewhere; I'm not sure. I'll see if I can update the endpoint to the wakatime one later this week. Thanks for the help!

**Comment by ghost** [comment]
Alright, I tested the compat endpoint and adapted the data format to the report. Also modified the config a bit, so it is a bit more flexible. Should work with Wakatime as well, but I don't have an account to test with. Basic setup should be as simple now as:
```bash
wget -O ~/.wakamonthrc https://raw.githubusercontent.com/bitstillery/wakamonth/main/.wakamonthrc.example
npx @bitstillery/wakamonth -y 2024 -m 1 report
```

**Reply by muety** [reply]
Very cool! :raised_hands: Only a few more minor remarks:
- There is a double-quote missing in the default config file
- I think it would be convenient if your could pass the project as a command-line argument, so you wouldn't have to modify the config file every time you request stats for a different project
- Initial setup would be even more convenient if the API URL and -key were automatically read from the `.wakatime.cfg` file

**Reply by ghost** [reply]
Thanks for the feedback! All valid and useful changes. Haven't thought about reusing an existing .wakatime.cfg yet! This would probably be an init command. I'll move the project to commandline. Made a start to support multiple outputs(stdout/xlsx). Not sure whether other formats might be useful.

**Comment by ghost** [comment]
Small blogpost about the whole process: https://bitstillery.com/2024/01/09/wakamonth-hours-reporting-tool/

**Reply by muety** [reply]
Awesome! I added your blog post to the [community contributions](https://github.com/muety/wakapi#-community-contributions) section.

**Comment by ghost** [comment]
I applied the wakatime.cfg changes, but it seems that it still needs a backend config parameter?
In my wakatime.cfg, I have a Wakapi endpoint like:
```
# Wakapi
api_url = https://wakapi.<DOMAIN>/api
# Wakatime
api_url=https://api.wakatime.com/api
```
The endpoints are:
```
# Wakapi
/compat/wakatime/v1/users/${user.id}/summaries
# Wakatime
/v1/users/${user.id}/summaries
```
Should I use `api_url = https://wakapi.<DOMAIN>/api/compat/wakatime` in wakatime.cfg instead of `https://wakapi.<DOMAIN>/api`? In that case the endpoints are the same, and I can loose the backend parameter as well.

**Reply by muety** [reply]
> Should I use `api_url = https://wakapi.<DOMAIN>/api/compat/wakatime` in wakatime.cfg instead of `https://wakapi.<DOMAIN>/api`?

Yes, exactly.

**Reply by ghost** [reply]
I just checked, but after I did update the `api_url` in wakatime.cfg to api/compat/wakatime, I didn't receive any data anymore (using the vscode plugin). Switching back to /api restores this. Not sure what went wrong; I'll try to debug it later. By the way, I moved further work on Wakamonth to this repo (https://codeberg.org/garage44/wakamonth)
